File: neetcode150/15_greedy/1899_merge_triplets_to_form_target_triplet.cpp
/*
 * =====================================================================
 * LeetCode 1899 : Merge Triplets to Form Target Triplet             (Medium)
 * https://leetcode.com/problems/merge-triplets-to-form-target-triplet/
 * Category   : Greedy
 * ---------------------------------------------------------------------
 * PROBLEM
 *   Merge (component-wise MAX) any subset of triplets; can the result
 *   equal the target triplet exactly?
 * ---------------------------------------------------------------------
 * INTUITION
 *   A triplet is USEFUL only if no component exceeds the target (a
 *   component > target can never be reduced). Merging all useful
 *   triplets is optimal: max is monotone, and the union of maxima
 *   only helps.
 * ---------------------------------------------------------------------
 * APPROACH 1 · Greedy filter + merge (best)
 *   1. Keep triplets with all three components ≤ target.
 *   2. Merge them; the running max equals target component-wise iff
 *      a solution exists.
 * APPROACH 2 · Coordinate tracking (equivalent, in-place)
 *   Track (a,b,c) as the best-so-far coordinate-wise maximum of the
 *   admissible triplets — identical math, less memory.
 * ---------------------------------------------------------------------
 * DEEP DIVE · Why "merge all useful" is optimal
 *   max(a,b) ≥ a and ≥ b: merging never destroys components. So if
 *   ANY subset can reach the target, the FULL set of admissible
 *   triplets reaches it too (it dominates every candidate subset
 *   coordinate-wise). Conversely, a triplet that exceeds the target
 *   in some coordinate poisons every merge containing it — it must be
 *   excluded entirely. This dichotomy (useful ⇔ admissible) makes the
 *   greedy trivially optimal: there is literally nothing to choose.
 * ---------------------------------------------------------------------
 * TIME COMPLEXITY : O(n) — single pass.
 * MEMORY COMPLEXITY: O(1) (A2) / O(n) (A1 storing the filter).
 * =====================================================================
 */
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <sys/resource.h>

class Solution_1 {
public:
    bool mergeTriplets(std::vector<std::vector<int>>& ts, std::vector<int>& target) {
        std::vector<std::vector<int>> ok;
        for (auto& t : ts)
            if (t[0] <= target[0] && t[1] <= target[1] && t[2] <= target[2])
                ok.push_back(t);
        std::vector<int> acc(3, 0);
        for (auto& t : ok)
            for (int i = 0; i < 3; ++i) acc[i] = std::max(acc[i], t[i]);
        return acc == target;
    }
};

class Solution_2 {
public:
    bool mergeTriplets(std::vector<std::vector<int>>& ts, std::vector<int>& target) {
        int a = 0, b = 0, c = 0;
        for (auto& t : ts)
            if (t[0] <= target[0] && t[1] <= target[1] && t[2] <= target[2]) {
                a = std::max(a, t[0]);
                b = std::max(b, t[1]);
                c = std::max(c, t[2]);
            }
        return a == target[0] && b == target[1] && c == target[2];
    }
};

static long long nowUs() {
    return std::chrono::duration_cast<std::chrono::microseconds>(
        std::chrono::steady_clock::now().time_since_epoch()).count();
}
static long memKb() { struct rusage r{}; getrusage(RUSAGE_SELF, &r); return r.ru_maxrss; }

int main() {
    struct Case { std::vector<std::vector<int>> ts; std::vector<int> t; bool want; };
    const Case cases[] = {
        {{{2, 5, 3}, {1, 8, 4}, {1, 7, 5}}, {2, 7, 5}, true},
        {{{2, 3, 4}, {1, 2, 5}, {3, 2, 2}}, {2, 3, 5}, true},
        {{{1, 1, 1}}, {1, 1, 2}, false},
        {{{3, 5, 3}}, {3, 5, 3}, true}
    };
    const long mem0 = memKb();
    auto t0 = nowUs();
    bool allOk = true;
    for (const auto& c : cases) {
        auto ts1 = c.ts, ts2 = c.ts;
        auto t1v = c.t, t2v = c.t;
        bool r1 = Solution_1().mergeTriplets(ts1, t1v);
        bool r2 = Solution_2().mergeTriplets(ts2, t2v);
        bool ok = r1 == c.want && r2 == c.want;
        allOk = allOk && ok;
        std::cout << "→ A1=" << (r1 ? "T" : "F") << " A2=" << (r2 ? "T" : "F")
                  << " (want " << (c.want ? "T" : "F") << ") "
                  << (ok ? "PASS" : "FAIL") << "\n";
    }
    auto t1 = nowUs();
    std::cout << (allOk ? "PASS : all cases" : "FAIL") << " | time: " << (t1 - t0)
              << " us | mem: " << (memKb() - mem0) << " KB\n";
    return 0;
}

File: neetcode150/15_greedy/1899_merge_triplets_to_form_target_triplet.py
"""
=====================================================================
LeetCode 1899 : Merge Triplets to Form Target Triplet             (Medium)
https://leetcode.com/problems/merge-triplets-to-form-target-triplet/
Category   : Greedy
---------------------------------------------------------------------
PROBLEM
    Merge (coordinate-wise max) any subset of triplets; can the
    result equal the target exactly?
---------------------------------------------------------------------
INTUITION
    Triplets with a coordinate > target poison every merge (max never
    shrinks). Merge ALL admissible triplets — max is monotone, so the
    full set dominates any candidate subset.
---------------------------------------------------------------------
APPROACH 1 · Filter then merge
    Keep triplets with all coordinates ≤ target; coordinate-wise max;
    compare to target.
APPROACH 2 · Running best (in-place)
    Track (a,b,c) best-so-far over admissible triplets; identical
    math, O(1) state.
---------------------------------------------------------------------
DEEP DIVE · Why "merge everything admissible" is optimal
    Since max(a,b) ≥ a, merging never destroys a coordinate. If ANY
    subset reaches the target, the full admissible set reaches it too
    (it dominates every subset coordinate-wise). And inadmissible
    triplets can't help — a coordinate beyond target can never come
    back down. Nothing to choose ⇒ greedy trivially optimal.
---------------------------------------------------------------------
TIME COMPLEXITY : O(n)
MEMORY COMPLEXITY: O(1)
=====================================================================
"""


class Solution_1:
    def mergeTriplets(self, triplets, target):
        ok = [t for t in triplets
              if all(t[i] <= target[i] for i in range(3))]
        acc = [0, 0, 0]
        for t in ok:
            for i in range(3):
                acc[i] = max(acc[i], t[i])
        return acc == target


class Solution_2:
    def mergeTriplets(self, triplets, target):
        a = b = c = 0
        for t in triplets:
            if t[0] <= target[0] and t[1] <= target[1] and t[2] <= target[2]:
                a = max(a, t[0])
                b = max(b, t[1])
                c = max(c, t[2])
        return [a, b, c] == target


if __name__ == "__main__":
    import time, tracemalloc
    cases = [([[2, 5, 3], [1, 8, 4], [1, 7, 5]], [2, 7, 5], True),
             ([[2, 3, 4], [1, 2, 5], [3, 2, 2]], [2, 3, 5], True),
             ([[1, 1, 1]], [1, 1, 2], False),
             ([[3, 5, 3]], [3, 5, 3], True)]
    tracemalloc.start()
    t0 = time.perf_counter()
    ok = True
    for ts, t, want in cases:
        r1, r2 = Solution_1().mergeTriplets(ts, t), Solution_2().mergeTriplets(ts, t)
        ok &= r1 == r2 == want
        print(f"-> A1={r1} A2={r2} (want {want}) {'PASS' if r1 == want and r2 == want else 'FAIL'}")
    dt = time.perf_counter() - t0
    mem = tracemalloc.get_traced_memory()[1] / 1024
    tracemalloc.stop()
    print(f"{'PASS : all cases' if ok else 'FAIL'} | time: {dt*1e6:.0f} us | mem: {mem:.0f} KB")
